### DISCUSSION TITLE
Graphql custom route

### DISCUSSION THREAD

**Discussion by guladima** [payloadcms/payload#2022] [Q&A] [answered]
Is it possible to change the route generation mechanism for graphql? I don't want my graphql url to start with `/api`.
https://github.com/payloadcms/payload/blob/4a49640c3fc22ee74a2666cb90588b03de68644d/src/graphql/initPlayground.ts#L4-L15

**Comment by DanRibbens** [comment]
That isn't currently configurable directly. I haven't seen this asked before, but if this is a common thing we could consider how it might be done in the config.

Alternatively, it may be possible to disable Payload's own graphQL using `graphQL.disable: true` in your config and set it up separately in your app by importing payload's graphql handler and adding it to your express router under your own path.

Payload sets up routes for the playground and graphql endpoints here:
https://github.com/payloadcms/payload/blob/4a49640c3fc22ee74a2666cb90588b03de68644d/src/initHTTP.ts#L56-L70

Disabling graphql will also skip registering of the schema, which you'll want to call also. I believe it could be done onInit in which you could pass it `payload` instead of `this`.
https://github.com/payloadcms/payload/blob/4a49640c3fc22ee74a2666cb90588b03de68644d/src/payload.ts#L194-L196

Since we don't export some of these files directly, you'll have to import them from 'payload/dist'.

I haven't tried to do this and honestly, it probably isn't worth the effort unless this is a hard requirement for your project.

I hope this helps, cheers!

**Comment by guladima** [comment, marked as answer]
Finally i solve my problem.

Here's my working solution:

### server.ts
```typescript
import express, { NextFunction, Response } from "express";
import payload from "payload";
import graphQLPlayground from "graphql-playground-middleware-express";
import registerSchema from "payload/dist/graphql/registerSchema";
import identifyAPI from "payload/dist/express/middleware/identifyAPI";
import graphQLHandler from "payload/dist/graphql/graphQLHandler";
import { PayloadRequest } from "payload/types";

require("dotenv").config();
const app = express();

const start = async () => {
  await payload.init({
    secret: process.env.PAYLOAD_SECRET,
    mongoURL: process.env.MONGODB_URI,
    express: app,
    onInit: async () => {
      payload.logger.info(`Payload Admin URL: ${payload.getAdminURL()}`);
      payload.logger.info(`Payload API URL: ${payload.getAPIURL()}`);
    },
  });

  app.use(
    payload.config.routes.graphQL,
    (req, res, next) => {
      if (req.method === "OPTIONS") {
        res.sendStatus(204);
      } else {
        next();
      }
    },
    (
      req: PayloadRequest & { payload: { schemaRegistered?: boolean } },
      _res: Response,
      next: NextFunction
    ) => {
      if (!req.payload.schemaRegistered) {
        registerSchema(req.payload);
        req.payload.schemaRegistered = true;
      }

      next();
    },
    identifyAPI("GraphQL"),
    (req: PayloadRequest, res: Response) => graphQLHandler(req, res)(req, res)
  );

  app.get(
    payload.config.routes.graphQLPlayground,
    graphQLPlayground({
      endpoint: payload.config.routes.graphQL,
      settings: {
        "request.credentials": "include",
      },
    })
  );

  app.listen(3000);
};

start();
```

### payload.config.ts
```typescript
routes: {
  admin: "/admin",
  api: "/api",
  graphQL: "/graphql",
  graphQLPlayground: "/graphql-playground",
},
graphQL: {
  disable: true,
  schemaOutputFile: path.resolve(__dirname, "generated-schema.graphql"),
},
```
